kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,700 @@
+      return answers, nil, ttl
+    end
+  end
+
+  if opts.cache_only then
+    return CACHE_ONLY_ANSWERS, nil, -1
+  end
+
+  local answers, err, ttl = resolve_query(self, name, qtype, tries)
+
+  if answers and answers.errcode == NAME_ERROR_CODE then
+    return nil  -- empty record for shm_miss cache
+  end
+
+  return answers, err, ttl
+end
+
+
+local function detect_recursion(opts, key)
+  local rn = opts.resolved_names
+  if not rn then
+    rn = {}
+    opts.resolved_names = rn
+  end
+  local detected = rn[key]
+  rn[key] = true
+  return detected
+end
+
+
+local function resolve_name_type(self, name, qtype, opts, tries)
+  local key = name .. ":" .. qtype
+
+  stats_init(self.stats, key)
+
+  if detect_recursion(opts, key) then
+    stats_count(self.stats, key, "fail_recur")
+    return nil, "recursion detected for name: " .. key
+  end
+
+  local answers, err, hit_level = self.cache:get(key, nil,
+                                                 resolve_name_type_callback,
+                                                 self, name, qtype, opts, tries)
+  -- check for runtime errors in the callback
+  if err and err:sub(1, 8) == "callback" then
+    log(ALERT, err)
+  end
+
+  -- restore the nil value in mlcache shm_miss to "name error" answers
+  if not answers and not err then
+    answers = NAME_ERROR_ANSWERS
+  end
+
+  local ctx = ngx.ctx
+  if ctx and ctx.has_timing then
+    req_dyn_hook_run_hooks("timing", "dns:cache_lookup",
+                           (hit_level and hit_level < HIT_L3))
+  end
+
+  -- hit L1 lru or L2 shm
+  if hit_level and hit_level < HIT_L3 then
+    stats_count(self.stats, key, hitstrs[hit_level])
+  end
+
+  if err or answers.errcode then
+    if not err then
+      local src = answers.errcode < CACHE_ONLY_EC and "server" or "client"
+      err = ("dns %s error: %s %s"):format(src, answers.errcode, answers.errstr)
+    end
+    table_insert(tries, { name .. ":" .. typstrs[qtype], err })
+  end
+
+  return answers, err
+end
+
+
+local function get_search_types(self, name, qtype)
+  local input_types = qtype and { qtype } or self.search_types
+  local checked_types = {}
+  local types = {}
+
+  for _, qtype in ipairs(input_types) do
+    if qtype == TYPE_LAST then
+      qtype = get_last_type(self.cache, name)
+    end
+    if qtype and not checked_types[qtype] then
+      table_insert(types, qtype)
+      checked_types[qtype] = true
+    end
+  end
+
+  return types
+end
+
+
+local function check_and_get_ip_answers(name)
+  if name:match("^%d+%.%d+%.%d+%.%d+$") then  -- IPv4
+    return {{ name = name, class = 1, type = TYPE_A, address = name }}
+  end
+
+  if name:match(":") then                     -- IPv6
+    return {{ name = name, class = 1, type = TYPE_AAAA, address = ipv6_bracket(name) }}
+  end
+
+  return nil
+end
+
+
+local function resolve_names_and_types(self, name, opts, tries)
+  local answers = check_and_get_ip_answers(name)
+  if answers then
+    answers.ttl = 10 * 365 * 24 * 60 * 60
```

Comment:
Should we define a constant for this ttl (`10 * 365 * 24 * 60 * 60`)?

Reply:
yes